### ISSUE TITLE
[Feature Request] Skip creating child token if token used is itself an orphan

### ISSUE THREAD

**Issue by mjtice** [hashicorp/terraform-provider-vault#550]
Right now the vault provider creates a child token with a reduced TTL and uses that token to access vault.(1)  The problem with this is that it requires the end user to use a non-orphaned token in the `token` argument.  This is because by their nature, non-token based authentication methods don't have a parent token(2).

I propose that terraform-provider-vault skips the child token creation if the token used to authenticate is an orphan token.

What do you think (I'm looking at you @tyrannosaurus-becks! :) )?  

1: https://github.com/terraform-providers/terraform-provider-vault/blob/master/vault/provider.go#L597
2: https://groups.google.com/d/msg/vault-tool/7Mq0t53mszA/VBCJGz1uAAAJ

**Comment by tstoermer** [comment]
Actually, I like the usage of a child token in the terraform run making sure everything is cleaned up and has a scoped lifetime.

In addition I would like to enable the child token with referencing information for auditing, see https://github.com/terraform-providers/terraform-provider-vault/pull/594

Maybe the creation of a child token should be controlled by a provider argument?

**Comment by tyrannosaurus-becks** [comment]
@mjtice I think that @tstoermer makes some good points. Also, there's a code note [here](https://github.com/terraform-providers/terraform-provider-vault/blob/master/vault/provider.go#L639-L650) regarding why that approach was taken.

What auth method are you using to obtain the token you'd like to use for running Vault? Also, can you provide more information about your use case that leads to your request for this change?

Thank you! For now I'm going to mark the associated PR as one not to be merged until we've reached a conclusion here.

**Comment by mjtice** [comment]
Hi @tyrannosaurus-becks 
Sure, I understand why the child token approach was taken.   Here's my use case:

1. We use Stackstorm ([and the associated terraform pack](https://github.com/StackStorm-Exchange/stackstorm-terraform) automate the creation and destruction of resources.
2. We've configured a local vault agent on this host and the agent is authenticated via approle method.
3. Using the token from the agent, I stuff it into my workflow (and therefor into my terraform.tfvars file) where it's consumed by the vault provider.
4. Because this token is an orphan and unable to create its own child token the terraform operations fail.

I'm not sure of the points @tstoermer is making - it sounds more like a feature request than an argument against what I'm proposing?  Maybe we take a route similar to what @tstoermer suggests (making the child token an argument) - but the default is to continue with what is currently done and an the function to skip creating a child is explicitly requested?

Maybe there's a way to do what I want and I'm just not seeing it?   I can always create a non-orphaned token and stick it in the KV store for retrieval - but it just ends up being something I have to manage (due to the TTL of the token).

Let me know what you think (I'm still lost in test-creation land for that PR anyway).

**Comment by tstoermer** [comment]
Hi @mjtice 
thanks for providing more details of your use case. I think your issue is not related to the fact if a token is an orphan or non-orphan one. I assume you are facing the following error message, when executing terraform with the AppRole orphan token:
`Error: failed to create limited child token: Error making API request.`

Please check the vault policy, that will be assigned to the AppRole token. It requires the token create capability:

```
path "auth/token/create" {
  capabilities = ["update"]
}
```

----

Regarding my proposal for the provider argument, would be something like:

```
provider "vault" {
  use_vault_child_token = false
}
```

It would be an optional argument with default value `true` and enables the terraform user to take  control of child token usage.

**Comment by mjtice** [comment]
I don't think that's right.  Correct me if I'm wrong on any of this:
1. The vault agent authenticates via approle
2. The token is pushed to the agent sink.
3. That token is an orphan token
4. By design, orphan tokens cannot create a child.

**Comment by tstoermer** [comment]
Hi @mjtice 
we are also using AppRole for our CI/CD job creating a token (orphan), that will be used with terraform managing Vault resources. We have no issues with the child token creation :wink:
Please verify the Vault policy assigned to the AppRole token - it must have proper capabilities. Also described in the [vault provider documentation](https://www.terraform.io/docs/providers/vault/index.html#provider-arguments) for the token argument.
Or do I overlook some special case?

**Comment by mjtice** [comment]
@tstoermer let me test this out some more.  I was able to create a child token from a orphan token (which I thought I couldn't...)

```
[(test) root@vault100.test. ~]# echo 'path "*" {
>   capabilities = ["create", "read", "update", "delete", "list", "sudo"]
> }' | vault policy write admin -
Success! Uploaded policy: admin
```

```
[(test) root@vault100.test. ~]# vault write auth/userpass/users/mjtice password=mysecret policies=admin
Success! Data written to: auth/userpass/users/mjtice
```
```
root@pc-dl28589:~# export VAULT_ADDR=http://pipeline.vault.service.pc1.consul.test.:8200
root@pc-dl28589:~# vault login -method=userpass username=mjtice password=mysecret
Success! You are now authenticated. The token information displayed below
is already stored in the token helper. You do NOT need to run "vault login"
again. Future Vault requests will automatically use this token.

Key                    Value
---                    -----
token                  s.90ZrYGBj2AGhRXTOrrniz4HQ
token_accessor         BhAC2s2Tu2czeaiijrb3ndYn
token_duration         768h
token_renewable        true
token_policies         ["admin" "default"]
identity_policies      []
policies               ["admin" "default"]
token_meta_username    mjtice
```

```
root@pc-dl28589:~# vault token lookup s.90ZrYGBj2AGhRXTOrrniz4HQ
Key                 Value
---                 -----
accessor            BhAC2s2Tu2czeaiijrb3ndYn
creation_time       1574180959
creation_ttl        768h
display_name        userpass-mjtice
entity_id           5a12f788-1e15-37cc-277d-6efcbd2cbc75
expire_time         2019-12-21T09:29:19.978366452-07:00
explicit_max_ttl    0s
id                  s.90ZrYGBj2AGhRXTOrrniz4HQ
issue_time          2019-11-19T09:29:19.978364563-07:00
meta                map[username:mjtice]
num_uses            0
orphan              true
path                auth/userpass/login/mjtice
policies            [admin default]
renewable           true
ttl                 767h59m48s
type                service

```
```
root@pc-dl28589:~# vault token create
Key                  Value
---                  -----
token                s.GPLjhkzHI7ktO97bFrknxKpE
token_accessor       fhWSrJVW1r65y9v1PVAcrtqq
token_duration       768h
token_renewable      true
token_policies       ["admin" "default"]
identity_policies    []
policies             ["admin" "default"]
```

**Comment by tstraley** [comment]
#775 was merged - please see PR for details.

When an upcoming release includes that change, you will be able to configure the vault provider to skip creation of the child token for situations like these where you are passing in an orphaned token. This may be sufficient to close this request.

```
provider "vault" {
  skip_child_token = true
}
```

**Comment by benashz** [comment]
Closed with #775
